### ISSUE TITLE
sometimes filter_lua converts double -> integer

### ISSUE THREAD

**Issue by nokute78** [fluent/fluent-bit#641]
When filter_lua converts Lua table to messagepack, some double values are converted to integer.

nop.lua
```lua
function nop(tag, timestamp, record)
     new_record = record
     return 1, timestamp, new_record
end
```

Configuration file.
Dummy generates 3 double values.
```python
[INPUT]
    Name dummy
    Tag  dummy

    dummy {"double1":0.9, "double2":1.0, "double3":1.1}

[FILTER]
    Name   lua
    Match  *
    script ./nop.lua
    call   nop

[OUTPUT]
    Name  stdout
    Match *
```

Output is 
```
[taka@localhost build]$ bin/fluent-bit -c dummy.conf 
Fluent-Bit v0.14.0
Copyright (C) Treasure Data

[2018/06/15 20:45:21] [ info] [engine] started (pid=15286)
[0] dummy: [1529063122.001215457, {"double2"=>1, "double1"=>0.900000, "double3"=>1.100000}]
[1] dummy: [1529063123.000427484, {"double2"=>1, "double1"=>0.900000, "double3"=>1.100000}]
[2] dummy: [1529063124.000121831, {"double2"=>1, "double1"=>0.900000, "double3"=>1.100000}]
[3] dummy: [1529063125.002114772, {"double2"=>1, "double1"=>0.900000, "double3"=>1.100000}]
```

In this case `double2` is converted integer.

This conversion is here.
https://github.com/fluent/fluent-bit/blob/master/plugins/filter_lua/lua.c#L179

**Comment by nokute78** [comment]
Note:  This article is to validate a lua number. (in Japanese)
https://qiita.com/iigura/items/4db51859fc49130d5f0c

**Comment by edsiper** [comment]
hmmm.. since Lua handle every number as a double, what would be the proper way to handle that case ?, should we use always double in the filter ?

**Comment by manuelluis** [comment]
@nokute78, the: https://qiita.com/iigura/items/4db51859fc49130d5f0c is talking about lua 5.3 that has integers.

With luajit, that is 5.1 and does not have integers, you get:
```
$ luajit
LuaJIT 2.1.0-beta3 -- Copyright (C) 2005-2017 Mike Pall. http://luajit.org/
JIT: ON SSE2 SSE3 SSE4.1 fold cse dce fwd dse narrow loop abc sink fuse
> print (1.2 % 1)           
0.2
> print (1.0 % 1)
0
> print (1 % 1)
0
```
you can not distinguish between 1.0 and 1, they are the same floating number.
Fluentbit is using luajit.

**Comment by edsiper** [comment]
@manuelluis do you think is better  in this case to pack that number always as a double ?

**Comment by nokute78** [comment]
@manuelluis Oops, thanks for checking.

@edsiper I think it isn't good to handle every number as a double.

If filter_lua modifies record depending on some conditions, 
the data types will be different from each cases.

conv.lua
```lua
function conv(tag, timestamp, record)
    if(record["cpu_p"] % 2 == 1) then
      new_record = record
      return 1, timestamp, new_record
    else
      return 0,0,0
    end
end
```

Configuration file
```python
[INPUT]
    Name cpu
    Tag  cpu

[FILTER]
    Name   lua
    Match  *
    script ./conv.lua
    call   conv

[OUTPUT]
    Name  stdout
    Match *
```

Output is
```
[taka@localhost build]$ bin/fluent-bit -c dummy.conf 
Fluent-Bit v0.14.0
Copyright (C) Treasure Data

[2018/06/16 08:26:50] [ info] [engine] started (pid=20307)
[0] cpu: [1529105211.001918905, {"cpu_p"=>0.000000, "user_p"=>0.000000, "system_p"=>0.000000, "cpu0.p_cpu"=>0.000000, "cpu0.p_user"=>0.000000, "cpu0.p_system"=>0.000000}]
[1] cpu: [1529105212.001008749, {"cpu0.p_cpu"=>1, "cpu0.p_user"=>1, "cpu_p"=>1, "cpu0.p_system"=>0, "system_p"=>0, "user_p"=>1}]
[2] cpu: [1529105213.004737854, {"cpu0.p_cpu"=>1, "cpu0.p_user"=>0, "cpu_p"=>1, "cpu0.p_system"=>1, "system_p"=>1, "user_p"=>0}]
[3] cpu: [1529105214.002930926, {"cpu_p"=>0.000000, "user_p"=>0.000000, "system_p"=>0.000000, "cpu0.p_cpu"=>0.000000, "cpu0.p_user"=>0.000000, "cpu0.p_system"=>0.000000}]
```

**Comment by manuelluis** [comment]
With #643 we can distinguish between integers and doubles.

**Comment by edsiper** [comment]
fixed by #643

**Comment by nokute78** [comment]
#643 is reverted. re-open this issue.

**Comment by edsiper** [comment]
Just for the record:

- we upgraded to LuaJIT 2.1.0-beta3 (mostly to support different architectures)
- at the moment considering also using vanilla Lua if portability to different OS/archs could be an issue

**Comment by edsiper** [comment]
FYI: addressed on c8f55c8b7850ca07c51002cfa2040642111eee18
